test/erc20_gateway/erc20_cogateway/setup.js

```diff
@@ -0,0 +1,109 @@
+// Copyright 2020 OpenST Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+const BN = require('bn.js');
+const Utils = require('../../test_lib/utils.js');
+const { AccountProvider } = require('../../test_lib/utils.js');
+const ConsensusGatewayUtils = require('../../consensus-gateway/utils');
+
+const ERC20Cogateway = artifacts.require('ERC20CogatewayDouble');
+
+contract('ERC20Cogateway::setup', (accounts) => {
+  const accountProvider = new AccountProvider(accounts);
+  let erc20Cogateway;
+  let setupGenesisParams = {};
+
+  beforeEach(async () => {
+    erc20Cogateway = await ERC20Cogateway.new();
+    setupGenesisParams = {
+      genesisMetachainId: Utils.generateRandomMetachainId(),
+      genesisERC20Gateway: accountProvider.get(),
+      genesisStateRootProvider: accountProvider.get(),
+      genesisMaxStorageRootItems: new BN(100),
+      genesisOutboxStorageIndex: new BN(4),
+    };
+    await erc20Cogateway.setupGenesis(
+      setupGenesisParams.genesisMetachainId,
+      setupGenesisParams.genesisERC20Gateway,
+      setupGenesisParams.genesisStateRootProvider,
+      setupGenesisParams.genesisMaxStorageRootItems,
+      setupGenesisParams.genesisOutboxStorageIndex,
+    );
+  });
+
+  it('should successfully setup ERC20 cogateway contract', async () => {
+
+    await erc20Cogateway.setup();
+
+    const outboxStorageIndexInContract = await erc20Cogateway.outboxStorageIndex.call();
+    assert.isOk(
+      setupGenesisParams.genesisOutboxStorageIndex.eq(outboxStorageIndexInContract),
+      `Expected outbox storage index is ${setupGenesisParams.genesisOutboxStorageIndex}`
+      + `but got ${outboxStorageIndexInContract}`,
+    );
+
+    const stateRootProviderInContract = await erc20Cogateway.stateRootProvider.call();
+    assert.strictEqual(
+      stateRootProviderInContract,
+      setupGenesisParams.genesisStateRootProvider,
+      'Incorrect state root provider in contract',
```

Comment:
```suggestion
      'State root provider address is not set in the contract.',
```

Reply:
Similarly for the other tests.

Reply:
Good point.
I think for `inboundChannelIdentifier` and `outboundChannelIdentifier`, the assertion message can be `Invalid inbound channel identifier.` and `Invalid outbound channel identifier.` because these values are calculated in the contract and we are asserting against the calculated values.